synapse/handlers/auth.py

```diff
@@ -770,13 +778,128 @@ def _auth_dict_for_flows(
+
+        The consumed refresh token is considered invalid after the first use of the new access token or the new refresh token.
+
+        Args:
+            refresh_token: The token to consume.
+            valid_until_ms: The expiration timestamp of the new access token.
+
+        Returns:
+            A tuple containing the new access token and refresh token
+        """
+
+        # Verify the token signature first before looking up the token
+        if not self.verify_refresh_token(refresh_token):
+            raise SynapseError(400, "invalid refresh token")
+
+        existing_token = await self.store.lookup_refresh_token(refresh_token)
+        if existing_token is None:
+            raise SynapseError(400, "refresh token does not exist")
+
+        if (
+            existing_token.has_next_access_token_been_used
+            or existing_token.has_next_refresh_token_been_refreshed
+        ):
+            raise SynapseError(400, "refresh token isn't valid anymore")
+
+        (
+            new_refresh_token,
+            new_refresh_token_id,
+        ) = await self.get_refresh_token_for_user_id(
+            user_id=existing_token.user_id, device_id=existing_token.device_id
+        )
+        access_token = await self.get_access_token_for_user_id(
+            user_id=existing_token.user_id,
+            device_id=existing_token.device_id,
+            valid_until_ms=valid_until_ms,
+            refresh_token_id=new_refresh_token_id,
+        )
+        await self.store.replace_refresh_token(
+            existing_token.token_id, new_refresh_token_id
+        )
+        return access_token, new_refresh_token
+
+    def _generate_refresh_token(self) -> str:
+        """
+        Generates a random and signed refresh token.
+
+        The token is signed with the first signing key in config
+        and can verified with verify_refresh_token.
+        """
+        rand = random.randbytes(128)
+        signed = self.signing_keys[0].sign(rand)
+        return "%s.%s" % (
+            base64.urlsafe_b64encode(signed.signature).decode("ascii"),
+            base64.urlsafe_b64encode(signed.message).decode("ascii"),
+        )
+
+    def verify_refresh_token(self, token: str) -> bool:
+        """
+        Verifies a refresh token signature.
+
+        Args:
+            token: The refresh token to verify
+
+        Returns:
+            Whether the token is valid or not
+        """
+        parts = token.split(".", maxsplit=2)
+        if len(parts) != 2:
+            return False
+        try:
+            signature = base64.urlsafe_b64decode(parts[0])
+            message = base64.urlsafe_b64decode(parts[1])
+        except binascii.Error:
+            return False
+
+        for key in self.signing_keys:
+            try:
+                key.verify_key.verify(message, signature)
+                return True
+            except nacl.signing.BadSignatureError:
+                pass
```

Comment:
Please check here if you're OK with my approach here. Tokens are basically shaped like this: `{base64(sig)}.{base64(rand)}` where `rand` is 128 random bytes and `sig` is a Ed25519 (iirc) signature of `rand` signed with the first signing key in the config.
When checking the validity of refresh token, we first check its signature (with old, rotated keys as well) before reaching the DB

Reply:
This is not the case anymore, I had not seen #5588 when I did this. Tokens now have the same shape as access tokens. I'm checking their shape + CRC before reaching the DB though.

Reply:
yup the new approach looks good.